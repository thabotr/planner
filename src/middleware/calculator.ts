const firstElevenPseudoFibonacciNumbers = [0, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89];
function toSubjectiveEffortScore(percent?: number): number {
    const zeroToTen = (percent ?? 0)/10;
    const intergralIndex = Math.floor(zeroToTen);
    return firstElevenPseudoFibonacciNumbers[intergralIndex];
}

export { toSubjectiveEffortScore };