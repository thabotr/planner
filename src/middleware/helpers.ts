
function percentToIntegralIndex(percent?: number): number {
    const zeroToTen = (percent ?? 0) / 10;
    return Math.floor(zeroToTen);
}
const firstElevenPseudoFibonacciNumbers = [0, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89];
function toSubjectiveEffortScore(percent?: number): number {
    const intergralIndex = percentToIntegralIndex(percent);
    return firstElevenPseudoFibonacciNumbers[intergralIndex];
}

function verboseTimestamp(diffInMs: number): string {
    const diffSecs = Math.floor(diffInMs / 1_000);
    const diffMins = Math.floor(diffSecs / 60);
    const diffHrs = Math.floor(diffMins / 60);
    const diffDays = Math.floor(diffHrs / 24);
    if (diffDays > 0) {
        return `${diffDays}dy`;
    }
    if (diffHrs > 0) {
        return `${diffHrs}h`;
    }
    if (diffMins > 0) {
        return `${diffMins}m`;
    }
    return `${diffSecs}s`;
}

const getRandomTime = () => Math.round(Math.random() * 1_000 * 60 * 60 * 24);

type TaskType = {
    pES: number,    // mental effort score
    mES: number,    // mental effort score
    length: number,     // time investment in milliseconds
}

type AvailabilityType = TaskType & {
    from: number,       // start time in milliseconds
};

type ScheduleItemType = {
    timeslot: AvailabilityType & { id: number },
    tasks: TaskType[],
};

class Scheduler {
    add(tslot: AvailabilityType): number | null {
        const item: ScheduleItemType = {
            timeslot: { ...tslot, id: this.#getId() },
            tasks: [],
        };

        const listIsEmpty = this.#list.length === 0;
        if (listIsEmpty) {
            this.#list.push(item);
            return item.timeslot.id;
        }

        let left = 0;
        let right = this.#list.length - 1;

        let middle: number;
        while (right - left > 1) {
            middle = Math.floor((right - left) / 2);
            if (this.#overlapsWithItemAt(tslot, left)) { return null; }
            if (this.#goesBeforeItemAt(tslot, middle)) { right = middle; }
            else { left = middle; }
        }

        if (this.#overlapsWithItemAt(tslot, left)) { return null; }

        if (this.#goesBeforeItemAt(tslot, left)) { this.#list.unshift(item); }
        else { this.#list.push(item); }

        return item.timeslot.id;
    }
    schedule(task: TaskType): number | null {
        for (let tslotIndex = 0; tslotIndex < this.#list.length; ++tslotIndex) {
            const scheduleTasks = this.#list[tslotIndex].tasks;
            const allTasks = scheduleTasks.concat(task);
            const superTask = allTasks.reduce((prev, curr) => ({
                length: prev.length + curr.length,
                mES: prev.mES + curr.mES,
                pES: prev.pES + curr.pES,
            }));
            const timeslot = this.#list[tslotIndex].timeslot;
            const superTaskFitsIntoTimeslot = timeslot.length >= superTask.length &&
                timeslot.mES >= superTask.mES && timeslot.pES >= superTask.pES

            if (superTaskFitsIntoTimeslot) {
                this.#list[tslotIndex].tasks = allTasks;
                return tslotIndex;
            }

        }
        return null;
    }
    getScheduleOn(day: Date): ScheduleItemType[] {
        const dayMask: AvailabilityType = {
            from: day.getTime(),
            length: AvailabilityDS.ADayInMillis,
            mES: 0,
            pES: 0,
        };
        return this.#list.filter((_, index) => this.#overlapsWithItemAt(dayMask, index));
    }
    remove(timeslotId: number): TaskType[] | null {
        let foundTslot2Rmv = false;
        for (const index in this.#list) {
            const { timeslot } = this.#list[index];
            foundTslot2Rmv = timeslot.id === timeslotId;
            if (foundTslot2Rmv) {
                const indexOfTSlot2Rmv = Number.parseInt(index);
                const lastItemIndex = this.#list.length - 1;
                const tasks4Tslot2rmv = this.#list[indexOfTSlot2Rmv].tasks;
                this.#list[indexOfTSlot2Rmv] = this.#list[lastItemIndex];
                this.#list.pop();
                return tasks4Tslot2rmv;
            }
        }
        return null;
    }
    #goesBeforeItemAt(tslot: AvailabilityType, index: number): boolean {
        const { timeslot: tslotAtIndex } = this.#list[index];
        return tslot.from < tslotAtIndex.from;
    }
    #overlapsWithItemAt(newTslot: AvailabilityType, index: number): boolean {
        const { timeslot } = this.#list[index];
        const newTslotStartContainedInTslot = timeslot.from <= newTslot.from &&
            newTslot.from <= timeslot.from + timeslot.length;
        const newTslotEndContainedInTslot = timeslot.from <= newTslot.from + newTslot.length &&
            newTslot.from + newTslot.length <= timeslot.from + timeslot.length;
        const newTslotContainsTslot = newTslot.from <= timeslot.from &&
            timeslot.from <= newTslot.from + newTslot.length;
        return newTslotStartContainedInTslot || newTslotEndContainedInTslot ||
            newTslotContainsTslot;
    }
    #id = 0;
    #getId() { ++this.#id; return this.#id - 1; }
    #list: ScheduleItemType[] = [];
}

class AvailabilityDS {
    // requirements
    // can add availability
    // cannot add availability if it intersects with already existing availability
    // should be able to provide us with this conflicting availability

    // can remove availability - should return it along with its scheduled contents else null

    // can schedule contents over some given availability provided the availability has enough capacity
    // can get scheduled contents and what part of the availability they consume

    // can split availability

    // can merge two adjacent availabilities else return false

    // can get availability given a single day
    list: Array<AvailabilityType> = [];
    #advancedList: Array<{ availability: AvailabilityType, tasks: TaskType[] }> = [];

    addAvailability(item: AvailabilityType): AvailabilityType | null {
        const availabilityOverlapsWithItem = (av: AvailabilityType) => item.from >= av.from && item.from <= av.from + av.length;
        if (this.list.some(availabilityOverlapsWithItem)) {
            return null;
        }
        this.list.push(item);
        return item;
    }
    getAvailabilityForDay(day: Date): Array<AvailabilityType> {
        const dayStartTime = day.getTime();
        const dayEndTime = dayStartTime + AvailabilityDS.ADayInMillis;
        return this.list.filter(av => av.from >= dayStartTime && av.from < dayEndTime);
    }

    scheduleTask(task: TaskType): AvailabilityType | null {
        return this.list.find(av => av.length >= task.length && av.mES >= task.mES && av.pES >= task.pES) ?? null;
    }

    clearAllAvailability() {
        this.list = [];
    }

    static ADayInMillis = 1_000 * 60 * 60 * 24;
    static AnHourInMillis = 1_000 * 60 * 60;
    static AMinuteInMillis = 1_000 * 60;
}

export { toSubjectiveEffortScore, verboseTimestamp, getRandomTime, AvailabilityDS, Scheduler, type AvailabilityType, type TaskType };