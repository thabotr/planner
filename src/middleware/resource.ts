import type DescriptiveItemType from '../types/DescriptiveItemType';
class Resource {
    async getTasks(): Promise<Array<DescriptiveItemType>> {
        // make get request
        return [];
    }

    async createTask(task: DescriptiveItemType) {
        // make post request
    }
}

export {Resource};