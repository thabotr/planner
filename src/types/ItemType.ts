class ItemType {
    length: number;
    mES: number;
    pES: number;
    id: string;
    constructor(length: number, mES: number, pES: number, id: string) {
        this.length = length;
        this.id = id;
        this.mES = mES;
        this.pES = pES;
    }
};

export default ItemType;