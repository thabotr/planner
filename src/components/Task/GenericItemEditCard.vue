<template>
    <GenericCard :item="tempItem">
        <div class="flex-horizontal gapped-s">
            <v-chip density="compact">{{ itemType }}</v-chip>
            <slot name="header"></slot>
        </div>
        <slot></slot>
        <template #actions>
            <div class="flex-vertical justify-space-between all-actions">
                <v-btn aria-label="delete" @click="$emit('delete')" icon="mdi-delete-forever" density="compact"
                    flat></v-btn>
                <div class="flex-vertical safe-actions">
                    <v-btn aria-label="cancel" @click="$emit('cancel')" icon="mdi-close" density="compact" flat></v-btn>
                    <v-btn aria-label="save" @click="$emit('save', tempItem)" icon="mdi-content-save" density="compact"
                        flat></v-btn>
                </div>
            </div>
        </template>
        <template #controls>
            <div id="controls" class="flex-vertical">
                <div id="mes-control">
                    <v-btn icon="mdi-minus" aria-label="decrement mes" density="compact" flat
                        @click="updateES('mES', 'dec')"></v-btn>
                    <v-btn icon="mdi-plus" aria-label="increment mes" density="compact" flat
                        @click="updateES('mES', 'inc')"></v-btn>
                </div>
                <div id="pes-control">
                    <v-btn icon="mdi-minus" aria-label="decrement pes" density="compact" flat
                        @click="updateES('pES', 'dec')"></v-btn>
                    <v-btn icon="mdi-plus" aria-label="increment pes" density="compact" flat
                        @click="updateES('pES', 'inc')"></v-btn>
                </div>
                <div id="length-control">
                    <v-btn icon="mdi-undo" aria-label="decrement time" density="compact" flat
                        @click="updateLen('dec')"></v-btn>
                    <v-btn icon="mdi-fast-forward-5" aria-label="increment time" density="compact" flat
                        @click="updateLen('inc')"></v-btn>
                </div>
            </div>
        </template>
    </GenericCard>
</template>

<script lang="ts">
import GenericCard from './GenericCard.vue';
import ItemType from '@/types/ItemType';
import { TimeInMillis } from '@/middleware/helpers';

export default {
    props: {
        item: {
            type: ItemType,
            required: true,
        },
        itemType: {
            type: String,
            required: true,
        },
        lowerBound: ItemType,
    },
    data() {
        return {
            tempItem: new ItemType(this.item.length, this.item.mES, this.item.pES, this.item.id),
        };
    },
    watch: {
        lowerBound(lowerBound?: ItemType) {
            const length = Math.max(lowerBound?.length ?? this.item.length, this.item.length);
            const pES = Math.max(lowerBound?.pES ?? this.item.pES, this.item.pES);
            const mES = Math.max(lowerBound?.mES ?? this.item.mES, this.item.mES);
            this.tempItem = new ItemType(length, mES, pES, this.item.id);
        }
    },
    components: { GenericCard },
    methods: {
        updateLen(op: 'inc' | 'dec') {
            const fiveMinutes = 5 * TimeInMillis.Minute;
            const sixHours = 6 * TimeInMillis.Hour;
            const minTime = this.lowerBound?.length ?? 0;
            switch (op) {
                case 'inc':
                    const increased = this.tempItem.length + fiveMinutes;
                    const incBoundedByMax6hours = Math.min(increased, sixHours);
                    this.tempItem.length = incBoundedByMax6hours;
                    return;
                case 'dec':
                    const descreased = this.tempItem.length - fiveMinutes;
                    const decBounded = Math.max(descreased, fiveMinutes, minTime);
                    this.tempItem.length = decBounded;
            }
        },
        updateES(name: 'mES' | 'pES', op: 'inc' | 'dec') {
            switch (name) {
                case 'mES':
                    const minMES = Math.max(0, this.lowerBound?.mES ?? 0);
                    this.tempItem.mES = this.boundedSteppedUpdate(op, this.tempItem.mES, minMES);
                    return;
                default:
                    const minPES = Math.max(0, this.lowerBound?.pES ?? 0);
                    this.tempItem.pES = this.boundedSteppedUpdate(op, this.tempItem.pES, minPES);
                    return;
            }
        },
        boundedSteppedUpdate(type: 'inc' | 'dec', value: number, effortScoreMin: number): number {
            const effortScoreMax = 100;
            const effortScoreStep = 10;
            switch (type) {
                case 'inc':
                    return Math.min(value + effortScoreStep, effortScoreMax);
                default:
                    return Math.max(value - effortScoreStep, effortScoreMin);
            }
        }
    }
}
</script>

<style scoped>
.v-btn {
    background-color: transparent;
    color: var(--color-primary);
}

.all-actions {
    gap: 1rem;
    padding: 0.2rem;
}

.safe-actions {
    gap: 0.2rem;
}

#controls {
    gap: 1rem;
}
</style>