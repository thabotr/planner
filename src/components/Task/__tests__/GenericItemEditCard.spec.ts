import GenericItemEditCard from "@/components/Task/GenericItemEditCard.vue";
import { TimeInMillis, toSubjectiveEffortScore, verboseTimestamp } from "@/middleware/helpers";
import ItemType from "@/types/ItemType";
import { beforeEach, describe, expect, it } from "vitest";
import { CustomVueWrapper } from "../../__tests__/utils";

describe('GenericItemEditCard', () => {
    const defaultItem = new ItemType(30 * TimeInMillis.Minute, 50, 50, "id");
    const renderComponent = () =>
        new CustomVueWrapper(GenericItemEditCard, {
            props: {
                item: defaultItem,
                itemType: "itemTypeGoesHere",
            },
            slots: {
                default: 'Default',
                header: 'Some header content goes here',
            }
        });
    let wrapper: CustomVueWrapper;
    beforeEach(() => {
        wrapper = renderComponent();
    });
    it('renders correctly', () => {
        expect(wrapper.wrapper.html()).toMatchSnapshot();
    });
    const assertEmmitedEvent = async (name: string) => {
        expect(wrapper.wrapper.emitted()).not.toHaveProperty(name);
        await wrapper.click(name);
        expect(wrapper.wrapper.emitted()).toHaveProperty(name);
    };
    it("emits 'delete' event on click delete", async () => {
        await assertEmmitedEvent('delete');
    });
    it("emits 'cancel' event on click cancel", async () => {
        await assertEmmitedEvent('cancel');
    });
    const effortScoreStep = 10;
    it("increments mental effort points on click increment mes", async () => {
        const increasedMes = defaultItem.mES + effortScoreStep;
        const meSMessage = `${toSubjectiveEffortScore(increasedMes)} mental EP`;
        expect(wrapper.wrapper.html()).not.toContain(meSMessage)
        await wrapper.click('increment mes');
        expect(wrapper.wrapper.html()).toContain(meSMessage);
    });
    it("decrements mental effort points on click decrement mes", async () => {
        const decreasedMes = defaultItem.mES - effortScoreStep;
        const mesMessage = `${toSubjectiveEffortScore(decreasedMes)} mental EP`;
        expect(wrapper.wrapper.html()).not.toContain(mesMessage)
        await wrapper.click('decrement mes');
        expect(wrapper.wrapper.html()).toContain(mesMessage);
    });
    it("increments physical effort points on click increment pes", async () => {
        const increasedPes = defaultItem.pES + effortScoreStep;
        const pesMessage = `${toSubjectiveEffortScore(increasedPes)} physical EP`;
        expect(wrapper.wrapper.html()).not.toContain(pesMessage);
        await wrapper.click('increment pes');
        expect(wrapper.wrapper.html()).toContain(pesMessage);
    });
    it("decrements physical effort points on click decrement pes", async () => {
        const decreasedPes = defaultItem.pES - effortScoreStep;
        const pesMessage = `${toSubjectiveEffortScore(decreasedPes)} physical EP`;
        expect(wrapper.wrapper.html()).not.toContain(pesMessage);
        await wrapper.click('decrement pes');
        expect(wrapper.wrapper.html()).toContain(pesMessage);
    });
    const fiveMinutes = 5 * TimeInMillis.Minute;
    it("increments time by five minutes on click increment time", async () => {
        const increasedTime = defaultItem.length + fiveMinutes;
        const increasedTimeMsg = verboseTimestamp(increasedTime);
        expect(wrapper.wrapper.html()).not.toContain(increasedTimeMsg);
        await wrapper.click('increment time');
        expect(wrapper.wrapper.html()).toContain(increasedTimeMsg);
    });
    it("decrements time on click decrement time", async () => {
        const decreasedTime = defaultItem.length - fiveMinutes;
        const decreasedTimeMsg = verboseTimestamp(decreasedTime);
        expect(wrapper.wrapper.html()).not.toContain(decreasedTimeMsg);
        await wrapper.click('decrement time');
        expect(wrapper.wrapper.html()).toContain(decreasedTimeMsg);
    });
    it("emits 'save' event with the contents of the updated item on click save", async () => {
        expect(wrapper.wrapper.emitted()).not.toHaveProperty('save');
        const updatedItem = new ItemType(
            defaultItem.length + fiveMinutes,
            defaultItem.mES + effortScoreStep,
            defaultItem.pES + effortScoreStep,
            defaultItem.id,
        );
        await wrapper.click('increment time');
        await wrapper.click('increment mes');
        await wrapper.click('increment pes');
        await wrapper.click('save');
        
        expect(wrapper.wrapper.emitted()).toHaveProperty('save');
        const [saveEventData] = wrapper.wrapper.emitted()['save'];
        const itemFromSaveEvent = (saveEventData as ItemType[])[0];
        expect(itemFromSaveEvent).toStrictEqual(updatedItem);
    });
});