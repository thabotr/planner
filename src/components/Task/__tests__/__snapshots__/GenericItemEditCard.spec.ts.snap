// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GenericItemEditCard > renders correctly 1`] = `
"<div data-v-f39f1573=\\"\\" data-v-9b7145d6=\\"\\">
  <div data-v-f39f1573=\\"\\" class=\\"v-card v-theme--light v-card--density-default v-card--variant-elevated\\" id=\\"bg\\">
    <!---->
    <div class=\\"v-card__loader\\">
      <div class=\\"v-progress-linear v-theme--light\\" style=\\"top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
        <!---->
        <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
        <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
          <div class=\\"v-progress-linear__indeterminate\\">
            <div class=\\"v-progress-linear__indeterminate long\\"></div>
            <div class=\\"v-progress-linear__indeterminate short\\"></div>
          </div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
    <!---->
    <div data-v-f39f1573=\\"\\">
      <div data-v-f39f1573=\\"\\" class=\\"v-card v-theme--light v-card--density-default v-card--variant-elevated\\" id=\\"fg\\">
        <!---->
        <div class=\\"v-card__loader\\">
          <div class=\\"v-progress-linear v-theme--light\\" style=\\"top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
            <!---->
            <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
            <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
              <div class=\\"v-progress-linear__indeterminate\\">
                <div class=\\"v-progress-linear__indeterminate long\\"></div>
                <div class=\\"v-progress-linear__indeterminate short\\"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <!---->
        <div data-v-9b7145d6=\\"\\" class=\\"flex-horizontal gapped-s\\"><span data-v-9b7145d6=\\"\\" class=\\"v-chip v-theme--light v-chip--density-compact v-chip--size-default v-chip--variant-tonal\\" draggable=\\"false\\"><!----><span class=\\"v-chip__underlay\\"></span>
          <!---->
          <!---->itemTypeGoesHere
          <!---->
          <!----></span>Some header content goes here
        </div>Default<div data-v-f39f1573=\\"\\" class=\\"flex-horizontal\\">
          <div data-v-19720537=\\"\\" data-v-f39f1573=\\"\\" class=\\"flex-vertical justify-space-between\\">
            <div data-v-19720537=\\"\\" class=\\"flex-horizontal\\"><i data-v-19720537=\\"\\" class=\\"mdi-brain mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\" id=\\"i-mental\\"></i>8 mental EP </div>
            <div data-v-19720537=\\"\\" class=\\"flex-horizontal\\"><i data-v-19720537=\\"\\" class=\\"mdi-account-hard-hat mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\" id=\\"i-physical\\"></i>8 physical EP </div>
            <div data-v-19720537=\\"\\" class=\\"flex-horizontal\\"><i data-v-19720537=\\"\\" class=\\"mdi-timer mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\" id=\\"i-time\\"></i>30m</div>
          </div>
          <div data-v-9b7145d6=\\"\\" id=\\"controls\\" class=\\"flex-vertical\\">
            <div data-v-9b7145d6=\\"\\" id=\\"mes-control\\" class=\\"flex-horizontal\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"decrement mes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
                <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-minus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
                <!---->
                <!---->
              </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"increment mes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
                <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-plus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
                <!---->
                <!---->
              </button></div>
            <div data-v-9b7145d6=\\"\\" id=\\"pes-control\\" class=\\"flex-horizontal\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"decrement pes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
                <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-minus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
                <!---->
                <!---->
              </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"increment pes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
                <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-plus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
                <!---->
                <!---->
              </button></div>
            <div data-v-9b7145d6=\\"\\" id=\\"length-control\\" class=\\"flex-horizontal\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"decrement time\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
                <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-undo mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
                <!---->
                <!---->
              </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"increment time\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
                <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-fast-forward-5 mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
                <!---->
                <!---->
              </button></div>
          </div>
        </div>
        <!---->
        <!----><span class=\\"v-card__underlay\\"></span>
      </div>
    </div>
    <div data-v-f39f1573=\\"\\" id=\\"actions\\" class=\\"padded-s\\"><span data-v-f39f1573=\\"\\" class=\\"v-chip v-theme--light v-chip--density-compact v-chip--size-default v-chip--variant-tonal\\" draggable=\\"false\\"><!----><span class=\\"v-chip__underlay\\"></span>
      <!---->
      <!---->id
      <!---->
      <!----></span>
      <div data-v-9b7145d6=\\"\\" class=\\"flex-vertical all-actions justify-space-between\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"delete\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
          <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-delete-forever mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
          <!---->
          <!---->
        </button>
        <div data-v-9b7145d6=\\"\\" class=\\"flex-vertical safe-actions\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"cancel\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
            <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-close mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
            <!---->
            <!---->
          </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"save\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
            <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-content-save mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
            <!---->
            <!---->
          </button></div>
      </div>
    </div>
    <!---->
    <!----><span class=\\"v-card__underlay\\"></span>
  </div>
</div>"
`;
